src/Tes.Runner/Events/EventSink.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Threading.Channels;
+using Microsoft.Extensions.Logging;
+using Tes.Runner.Transfer;
+
+namespace Tes.Runner.Events
+{
+    public abstract class EventSink : IEventSink
+    {
+        const int StopWaitDurationInSeconds = 60;
+
+        private readonly Channel<EventMessage> events = Channel.CreateUnbounded<EventMessage>();
+        private readonly ILogger logger = PipelineLoggerFactory.Create<EventSink>();
+        private Task? eventHandlerTask;
+
+        public abstract Task HandleEventAsync(EventMessage eventMessage);
+
+        public async Task PublishEventAsync(EventMessage eventMessage)
+        {
+            await events.Writer.WriteAsync(eventMessage);
+        }
+
+        public void Start()
+        {
+            logger.LogDebug("Starting events processing handler");
+
+            eventHandlerTask = Task.Run(async () => await EventHandlerAsync());
+        }
+
+
+        public async Task StopAsync()
+        {
+            if (eventHandlerTask is null)
+            {
+                throw new InvalidOperationException("Event handler task is not running");
+            }
+
+            //close the channel to stop the event handler
+            events.Writer.Complete();
+
+            await eventHandlerTask.WaitAsync(TimeSpan.FromSeconds(StopWaitDurationInSeconds));
+        }
+
+        protected IDictionary<string, string> ToEventTag(EventMessage eventMessage)
+        {
+            return new Dictionary<string, string>
+            {
+                { "event_name", eventMessage.Name },
+                { "event_id", eventMessage.Id },
+                { "entity_type", eventMessage.EntityType },
+                { "task_id", eventMessage.EntityId },
+                { "workflow_id", eventMessage.CorrelationId },
+                //format date to ISO 8601, which is URL friendly
+                { "created", eventMessage.Created.ToString("yyyy-MM-ddTHH:mm:ssZ") }
+            };
+        }
+        private async Task EventHandlerAsync()
+        {
+            while (await events.Reader.WaitToReadAsync())
```

Comment:
could this ever throw?  would it be better to put all of this inside a while(true){try{}catch{} await Task.Delay(1000);} equivalent so an exception will never crash the method?

Reply:
I don't think it is necessary. I can't think of a realistic scenario where checking if the channel has data would throw. Also, once the channel reaches an end/error state, keep trying in a forced loop would not help. However, if you can think of a scenario this method can throw, let's discuss it and see what is the best way to mitigate it.

Reply:
Great just wanted to check.  Maybe any of these?

`ObjectDisposedException`: If the underlying channel is disposed of before or during the operation.
`ChannelCompletionException`: If the channel has been marked as complete with an exception, then waiting to read could propagate that exception.
`OperationCanceledException`: If the CancellationToken passed to it gets canceled.
Other unexpected exceptions: In the case of bugs, resource issues, or unexpected behavior, other exceptions might be thrown, although these are not typical for this operation.